python/src/transforms.cpp

```diff
@@ -593,7 +593,122 @@ class PyCheckpointedFun {
   nb::callable fun_;
 };
 
+class PyCustomVJPFun {
+ public:
+  PyCustomVJPFun(nb::callable fun) : fun_(std::move(fun)) {}
+  ~PyCustomVJPFun() {
+    nb::gil_scoped_acquire gil;
+
+    fun_.release().dec_ref();
+    if (vjp_fun_.has_value()) {
+      (*vjp_fun_).release().dec_ref();
+    }
+  }
+
+  struct InnerFunction {
+    nb::callable vjp_fun_;
+    nb::object input_structure_;
+    std::shared_ptr<nb::object> output_structure_;
+
+    InnerFunction(
+        nb::callable vjp_fun,
+        nb::object input_structure,
+        std::shared_ptr<nb::object> output_structure)
+        : vjp_fun_(std::move(vjp_fun)),
+          input_structure_(std::move(input_structure)),
+          output_structure_(std::move(output_structure)) {}
+    ~InnerFunction() {
+      nb::gil_scoped_acquire gil;
+
+      vjp_fun_.release().dec_ref();
+      input_structure_.release().dec_ref();
+      if (output_structure_.use_count() == 1) {
+        output_structure_->release().dec_ref();
+      }
+    }
+
+    std::vector<array> operator()(
```

Comment:
I'm slightly confused on the purpose of this `operator()` overload. Is it used for anything?

Reply:
Yeah, this is the `InnerFunction` implementation. Basically an old school lambda to pass to `custom_vjp` (C++ side) as the vjp function. The reason for its existence (same as in `checkpoint`) is to control the destruction of the captured (in this case wrapped) python variables.

Basically this exists only for its destructor, otherwise it would be a simple lambda passed in line 686.